tests/unit/test_tracing.c

```diff
@@ -1377,37 +1378,165 @@ SENTRY_TEST(set_trace_id_with_txn)
         = "2674eb52d5874b13b560236d6c79ce8a-a0f9fdf04f1a63df-1";
     const char *txn_trace_id = "2674eb52d5874b13b560236d6c79ce8a";
     const char *txn_parent_span_id = "a0f9fdf04f1a63df";
+    sentry_transaction_context_t *tx_ctx_scoped
+        = sentry_transaction_context_new("wow!", NULL);
+    sentry_transaction_context_update_from_header(
+        tx_ctx_scoped, "sentry-trace", trace_header);
+    sentry_transaction_t *tx_scoped
+        = sentry_transaction_start(tx_ctx_scoped, sentry_value_new_null());
+
+    // when no set_trace was called yet, the scoped transaction should apply
+    //  its trace/parent span ID (as set by update_from_header)
+    sentry_set_transaction_object(tx_scoped);
+    apply_scope_and_check_trace_context(
+        options, txn_trace_id, txn_parent_span_id);
+    sentry_transaction_finish(tx_scoped);
+
     sentry_transaction_context_t *tx_ctx
         = sentry_transaction_context_new("wow!", NULL);
     sentry_transaction_context_update_from_header(
         tx_ctx, "sentry-trace", trace_header);
     sentry_transaction_t *tx
         = sentry_transaction_start(tx_ctx, sentry_value_new_null());
+    sentry_span_t *span_child
+        = sentry_transaction_start_child(tx, "op", "desc");
+    sentry_span_t *span_grandchild
+        = sentry_span_start_child(span_child, "op_g", "desc_g");
 
     // set the direct trace first
     const char *direct_trace_id = "aaaabbbbccccddddeeeeffff00001111";
     const char *direct_parent_span_id = "f0f0f0f0f0f0f0f0";
     sentry_set_trace(direct_trace_id, direct_parent_span_id);
 
     // events should get that trace applied
+    // TODO reason why they should get that trace applied
+    //  (tx started before set_trace, but trace was set so events get that data)
     apply_scope_and_check_trace_context(
         options, direct_trace_id, direct_parent_span_id);
 
     // now set a scoped transaction
+    // this should have the set_trace data applied to the tx (tx gets scoped)
     sentry_set_transaction_object(tx);
 
-    // events should get the transaction's trace applied
+    // check if trace_id was passed down properly
+    // TODO maybe refactor these span checks into function
+    TEST_CHECK_STRING_EQUAL(
+        sentry_value_as_string(sentry_value_get_by_key(tx->inner, "trace_id")),
+        direct_trace_id);
+    // TODO we only set the transaction on the scope; how about the child spans?
+    //  only apply the trace IF they also get scoped? or should they now
+    //  inherit?
+    sentry_set_span(span_child);
+    TEST_CHECK_STRING_EQUAL(sentry_value_as_string(sentry_value_get_by_key(
+                                span_child->inner, "trace_id")),
+        direct_trace_id);
+    // TODO grandchild does not get scoped, so what should its trace_id be?
+    TEST_CHECK_STRING_NOT_EQUAL(sentry_value_as_string(sentry_value_get_by_key(
+                                    span_grandchild->inner, "trace_id")),
+        direct_trace_id);
```

Comment:
This assumes the 2nd option described above; child spans created before `set_trace` only inherit the `set_trace` values if they get scoped (and do not get it from scoping the root span)

Reply:
ditto.